packages/hub/services/gas-estimation.ts

```diff
@@ -0,0 +1,139 @@
+  cardpay = inject('cardpay');
+  ethersProvider = inject('ethers-provider', { as: 'ethersProvider' });
+  readonly gasLimitTTL = 86400000; //1 days
+
+  async estimate(params: GasEstimationParams) {
+    let prisma = await this.prismaManager.getClient();
+    let gasLimit = await prisma.gasEstimationResult.findFirst({
+      where: {
+        chainId: params.chainId,
+        scenario: params.scenario,
+        tokenAddress: params.tokenAddress,
+        gasTokenAddress: params.gasTokenAddress,
+      },
+    });
+
+    if (gasLimit && addMilliseconds(gasLimit.updatedAt, this.gasLimitTTL) > nowUtc()) {
+      return gasLimit;
+    }
+
+    let provider = this.ethersProvider.getInstance(params.chainId);
+    let signer = new Wallet(config.get('hubPrivateKey'));
+    let scheduledPaymentModule = await this.cardpay.getSDK('ScheduledPaymentModule', provider, signer);
+    let gas;
+    let tokenAmount = ethers.utils.parseUnits('1', 'gwei');
+    let gasTokenAmount = ethers.utils.parseUnits('0.01', 'gwei');
+    switch (params.scenario) {
+      case GasEstimationResultsScenarioEnum.create_safe_with_module:
+        gas = (
+          await scheduledPaymentModule.createSafeWithModuleAndGuardEstimation({ from: signer.address })
+        ).toNumber();
+        break;
+      case GasEstimationResultsScenarioEnum.execute_one_time_payment:
+        if (
+          !params.tokenAddress ||
+          params.tokenAddress === '' ||
+          !params.gasTokenAddress ||
+          params.gasTokenAddress === ''
+        ) {
+          throw Error(`tokenAddress and gasTokenAddress is required in ${params.scenario}`);
+        }
+        gas = await scheduledPaymentModule.estimateExecutionGas(
+          this.getHubSPModuleAddress(params.chainId),
+          params.tokenAddress,
+          tokenAmount.toString(),
+          signer.address,
+          gasTokenAmount.toString(),
+          params.gasTokenAddress,
+          'salt1',
+          gasTokenAmount.toString(),
+          Math.round(nowUtc().getTime() / 1000),
+          null,
+          null
+        );
+        break;
+      case GasEstimationResultsScenarioEnum.execute_recurring_payment:
+        if (
+          !params.tokenAddress ||
+          params.tokenAddress === '' ||
+          !params.gasTokenAddress ||
+          params.gasTokenAddress === ''
+        ) {
+          throw Error(`tokenAddress and gasTokenAddress is required in ${params.scenario}`);
+        }
+        gas = await scheduledPaymentModule.estimateExecutionGas(
+          this.getHubSPModuleAddress(params.chainId),
+          params.tokenAddress,
+          tokenAmount.toString(),
+          signer.address,
+          gasTokenAmount.toString(),
+          params.gasTokenAddress,
+          'salt1',
+          gasTokenAmount.toString(),
+          null,
+          28,
+          Math.round(addDays(nowUtc(), 30).getTime() / 1000)
+        );
+        break;
+    }
+
+    gasLimit = await prisma.gasEstimationResult.upsert({
```

Comment:
Is there any point in keeping old records? Should we delete the old when inserting a new one?

Reply:
Nevermind, I just realized it won't add new records, just update the existing one.